test/jdk/java/net/Socket/CloseOnFailureTest.java

```diff
@@ -0,0 +1,271 @@
+                    .findStaticVarHandle(Socket.class, "factory", SocketImplFactory.class);
+        } catch (NoSuchFieldException | IllegalAccessException error) {
+            throw new RuntimeException(error);
+        }
+    }
+
+    private static void withSocketImplFactory(SocketImplFactory newFactory, ThrowingRunnable runnable) throws Throwable {
+        SocketImplFactory oldFactory = (SocketImplFactory) SOCKET_IMPL_FACTORY_HANDLE.getAndSet(newFactory);
+        try {
+            runnable.run();
+        } finally {
+            SOCKET_IMPL_FACTORY_HANDLE.set(oldFactory);
+        }
+    }
+
+    private static final class MockSocketImpl extends ThrowingSocketImpl {
+
+        private final AtomicInteger closeInvocationCounter = new AtomicInteger(0);
+
+        private final Exception bindError;
+
+        private final Exception connectError;
+
+        private MockSocketImpl(Exception bindError, Exception connectError) {
+            this.bindError = bindError;
+            this.connectError = connectError;
+        }
+
+        @Override
+        protected void create(boolean stream) {}
+
+        @Override
+        protected void bind(InetAddress host, int port) throws IOException {
+            throwIfPresent(bindError);
+        }
+
+        @Override
+        protected void connect(SocketAddress address, int timeoutMillis) throws IOException {
+            throwIfPresent(connectError);
+        }
+
+        private void throwIfPresent(Exception connectError) throws IOException {
+            if (connectError != null) {
+                switch (connectError) {
+                    case IOException error -> throw error;
+                    case RuntimeException error -> throw error;
+                    default -> throw new IllegalStateException(
+                            "unknown connectError type: " + connectError.getClass().getCanonicalName());
+                }
+            }
+        }
+
+        @Override
+        protected void close() {
+            closeInvocationCounter.incrementAndGet();
+        }
+
+    }
+
+    @ParameterizedTest
+    @MethodSource("ctor_should_close_on_failures")
+    @SuppressWarnings("resource")
+    void ctor_should_close_on_failures(TestCase testCase) throws Throwable {
+
+        // Create a socket using the mock `SocketImpl` configured to fail
+        withSocketImplFactory(() -> testCase.socketImpl, () -> {
+
+            // Trigger the failure
+            Exception error = assertThrows(Exception.class, () -> {
+                InetAddress address = InetAddress.getLoopbackAddress();
+                // Address and port are mostly ineffective.
+                // They just need to be _valid enough_ to reach to the point where both `SocketImpl#bind()` and `SocketImpl#connect()` are invoked.
+                // Failure will be triggered by the injected `SocketImpl`.
+                new Socket(address, 0xDEAD, address, 0xBEEF);
```

Comment:
Using new Socket(address, 0xDEAD, address, 0xBEEF);       using hex as port number, a bit too clever!  

I think a constant would be preferable for us simple folk.

private static final int DEADSERVERPORT = 0xDEAD; // or  57005

I think it is preferable for the local port to use 0 i.e. an OS ephemeral allocation, as it is moot in this test context

One observation I’d make on this is that, when scanning the code and not carrying the details of the MockSocketImpl with you, one is inclined to think that this will intermittently lead to BindException and ConnectException as both 0xDEAD and 0xBEEF are in the ephemeral port range

Thus, a general question on the scope of testing. When considering the use of mock object in the call flows. Are these tests sufficient for the testing that the spec changes are upheld by the full default implementation ?

Reply:
Applied these suggestions in e37ee840b9681aa017f8653df68c69232a6b6c83.

> One observation I’d make on this is that, when scanning the code and not carrying the details of the MockSocketImpl with you, one is inclined to think that this will intermittently lead to BindException and ConnectException as both 0xDEAD and 0xBEEF are in the ephemeral port range

Agreed. That's why I documented the choice of address and port just above the line calling the ctor. The new code you suggested using variables instead of hardcoded constants should further help with this.

> Thus, a general question on the scope of testing. When considering the use of mock object in the call flows. Are these tests sufficient for the testing that the spec changes are upheld by the full default implementation?

That is a very good question! Judging from [the CSR](https://bugs.openjdk.org/browse/JDK-8344527), the specification changes concern:

1. `SocketImpl::connect()`  failures (the connection cannot be established or the timeout expires before the connection is established)
2. `Socket::connect(SocketAddress)` is passed an unresolved address

(1) is verified using `connectShouldCloseOnFailures()`, and (2) is verified using `connectShouldCloseOnUnresolvedAddress()`. If, at a later time, `Socket::connect()` gets modified to **not** use the `SocketImpl`, and invalidate the assumptions held in this test, associated test methods will fail, and force the author to take the CSR requirements into account, AFAICT.